### PR TITLE
feat: add pre check for rate limit

### PR DESCRIPTION
**Pull request by yjhmelody**
The outer result of `until_key_n_ready_with_jitter` should be checked before start service, it's static result relatively .
And its meaning is the cap is insufficient for `n`.

### CODE REVIEW

File: src/server.rs [hunk 1]
```diff
@@ -51,6 +51,12 @@ pub async fn build(config: Config) -> anyhow::Result<SubwayServerHandle> {
 
     let rpc_method_weights = MethodWeights::from_config(&config.rpcs.methods);
 
+    // pre-check stage
+    if let Some(r) = &rate_limit_builder {
```
Comment (ermalkaleci) [inline src/server.rs:55]: if you want to check this then just make sure no weight is greater than burst (ip or connection)

Comment (yjhmelody) [reply src/server.rs:55]: Yeah, The result is fixed if config is not changed, so we should do pre-check.

**Review by ermalkaleci — CHANGES_REQUESTED** [review]
you can simple verify config file

**Comment by ermalkaleci** [comment]
> > you can simple verify config file
> 
> This code is doing what you said.

no it's not. you're adding additional code that needs to be maintained and no necessary at all.
All you need to do is validate config file. `assert!(method.rate_limit_weight <= min(ip.burst, connection.burst))`

**Comment by ermalkaleci** [comment]
it can be done here https://github.com/AcalaNetwork/subway/blob/ac19fefb38a001a30c59922ad7a33b6f9cd2cdd7/src/config/mod.rs#L206C14-L206C22

**Comment by yjhmelody** [comment]
That's not true, you just assume `period_secs ==1s`

**Comment by yjhmelody** [comment]
Wonder the behavior design: why still continue the service when validation failed.

**Comment by ermalkaleci** [comment]
because it runs on it's own thread. it should work for any period

File: src/config/mod.rs [hunk 2]
```diff
@@ -208,6 +212,43 @@ pub async fn validate(config: &Config) -> Result<(), anyhow::Error> {
 
     // validate use garde::Validate
     config.validate(&())?;
+
+    if let Some(rate_limit) = config.extensions.rate_limit.as_ref() {
+        if let Some(ref rule) = rate_limit.ip {
+            let burst = NonZeroU32::new(rule.burst).unwrap();
```
Comment (xlc) [inline src/config/mod.rs:218]: use `ok_or` to return an error in case burst is zero

**Review by xlc — APPROVED** [review]
some tests will be nice

**Comment by ermalkaleci** [comment]
> All you need to do is validate config file. `assert!(method.rate_limit_weight <= min(ip.burst, connection.burst))`
In case you missed this

**Comment by yjhmelody** [comment]
> > All you need to do is validate config file. `assert!(method.rate_limit_weight <= min(ip.burst, connection.burst))`
> 
> @xlc @yjhmelody In case you missed this

Hi, you are right, I have simplify the code.

**Review by ermalkaleci — APPROVED** [review]

**Pull request merged**